crates/validation_common/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use serde::{Deserialize, Deserializer};
+use std::borrow::Borrow;
+use std::collections::HashSet;
+use std::ops::Deref;
+
+/// Digest generic in hash size `N`
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Digest<const N: usize>(pub [u8; N]);
+
+impl<'de, const N: usize> Deserialize<'de> for Digest<N> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let dig: String = Deserialize::deserialize(deserializer)?;
+        let dig = hex::decode(dig).map_err(|e| Error::custom(format!("invalid hex: {e}")))?;
+        let dig = dig.try_into().map_err(|v: Vec<_>| {
+            Error::custom(format!(
+                "expected digest to have length of {N}, got {}",
+                v.len()
+            ))
+        })?;
+        Ok(Digest(dig))
+    }
+}
+
+impl<const N: usize> AsRef<[u8; N]> for Digest<N> {
+    fn as_ref(&self) -> &[u8; N] {
+        &self.0
+    }
+}
+
+impl<const N: usize> Deref for Digest<N> {
+    type Target = [u8; N];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<const N: usize> Borrow<[u8; N]> for Digest<N> {
+    fn borrow(&self) -> &[u8; N] {
+        &self.0
+    }
+}
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize)]
+#[serde(deny_unknown_fields)]
+struct UnvalidatedMeasurements<const N: usize> {
+    #[serde(default, rename = "enarx_signer")]
+    signer: HashSet<Digest<N>>,
+
+    #[serde(default, rename = "enarx_hash")]
+    hash: HashSet<Digest<N>>,
+}
+
+impl<const N: usize> TryFrom<UnvalidatedMeasurements<N>> for Measurements<N> {
+    type Error = &'static str;
+
+    fn try_from(
+        UnvalidatedMeasurements { signer, hash }: UnvalidatedMeasurements<N>,
+    ) -> Result<Self, Self::Error> {
+        if signer.is_empty() && hash.is_empty() {
+            Err("one of `enarx_signer` or `enarx_hash` must be specified")
+        } else {
+            Ok(Self {
+                signer: HashSet::from_iter(signer.into_iter()),
```

Comment:
```suggestion
                signer,
```
See the patch I sent you